zdx/writer.go

```diff
@@ -28,102 +31,192 @@ import (
+		// encounted, just close the underlying storage object and return.
+		return w.iow.Close()
+	}
+	// First, close the parent if it exists (which will recursively close
+	// all the parents to the root) while leaving the base layer open.
+	for p := w.writer.parent; p != nil; p = p.parent {
+		if err := p.Close(); err != nil {
 			return err
 		}
 	}
-	if err := w.out.Flush(); err != nil {
+	// Also, close the frame of the base layer so we can copy the hierarchy
+	// to the base.  Note that sum of the sizes of the parents is much smaller
+	// than the base so this will go fast compared to the entire indexing job.
+	if err := w.writer.closeFrame(); err != nil {
+		return err
+	}
+	// The hierarchy is now flushed and closed.  Assemble the file into
+	// a single microindex and remove the temporary btree files.
+	if err := w.finalize(); err != nil {
+		w.writer.buffer.Close()
 		return err
 	}
-	if err := w.writer.Close(); err != nil {
+	// Finally, close the base layer.
+	return w.writer.buffer.Close()
+}
+
+func (w *Writer) finalize() error {
+	// First, collect up parent linkage into a slice so we can traverse
+	// top down...
+	base := w.writer
+	var layers []*indexWriter
+	for p := base.parent; p != nil; p = p.parent {
+		layers = append(layers, p)
+	}
+	// Now, copy each non-base file in top-down order to the base-layer object.
+	var sizes []int64
+	sizes = append(sizes, base.frameStart)
+	for k := len(layers) - 1; k >= 0; k-- {
+		// Copy the files in the reverse order so the root comes first.
+		// This will avoid backward seeks while looking up keys in the tree
+		// (except for the one backward seek to the base layer).
+		layer := layers[k]
+		size := layer.frameStart
+		sizes = append(sizes, size)
+		f, err := os.Open(layer.name)
+		if err != nil {
+			return err
+		}
+		n, err := io.Copy(base.buffer, f)
+		if err != nil {
+			f.Close()
+			return err
+		}
+		if n != size {
+			f.Close()
+			return fmt.Errorf("internal microindex error: index file size (%d) does not equal zng size (%d)", n, size)
+		}
+		if err := f.Close(); err != nil {
+			return err
+		}
+	}
+	// Finally, write the size records as the trailer of the microindex.
+	rec, err := newTrailerRecord(w.zctx, w.childField, w.frameThresh, sizes, w.keyType)
+	if err != nil {
 		return err
 	}
-	if w.parent != nil {
-		return w.parent.Close()
+	if err := base.zng.Write(rec); err != nil {
+		return err
 	}
+	base.zng.EndStream()
 	return nil
 }
 
-func (w *Writer) Write(rec *zng.Record) error {
-	offset := w.out.Position()
+func newIndexWriter(base *Writer, w io.WriteCloser, name string) (*indexWriter, error) {
+	writer := bufwriter.New(w)
+	return &indexWriter{
+		base:     base,
+		buffer:   writer,
+		name:     name,
+		zng:      zngio.NewWriter(writer, zio.WriterFlags{}),
+		frameEnd: int64(base.frameThresh),
+	}, nil
+}
+
+func (w *indexWriter) newParent() (*indexWriter, error) {
+	file, err := ioutil.TempFile(w.base.tmpdir, "*")
```

Comment:
NIt: We conventionally use "", which is equivalent in this context to "*".
```suggestion
	file, err := ioutil.TempFile(w.base.tmpdir, "")
```